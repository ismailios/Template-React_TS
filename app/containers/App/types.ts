import { ActionType } from 'typesafe-actions';
import * as actions from './actions';
import { Repo } from '../RepoListItem/types';

/* --- STATE --- */

interface AppState {
  readonly loading: boolean;
  readonly error?: object | boolean;
  readonly currentUser: string;
  readonly userData: UserData;
}

interface UserData {
  readonly repos?: Repo[];
}

/* --- ACTIONS --- */
type AppActions = ActionType<typeof actions>;

/* --- EXPORTS --- */

type ContainerState = AppState;
type ContainerActions = AppActions;

export { ContainerState, ContainerActions, UserData };
