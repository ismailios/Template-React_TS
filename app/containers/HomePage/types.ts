import { ActionType } from 'typesafe-actions';
import * as actions from './actions';

/* --- STATE --- */

interface HomeState {
  readonly username: string;
}

/* --- ACTIONS --- */
type AppActions = ActionType<typeof actions>;

/* --- EXPORTS --- */

type ContainerState = HomeState;
type ContainerActions = AppActions;

export { ContainerState, ContainerActions };
