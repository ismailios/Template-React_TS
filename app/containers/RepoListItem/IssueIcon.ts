import styled from 'styles/styled-components';

import NormalIssueIcon from 'components/IssueIcon';

const IssueIcon = styled(NormalIssueIcon)`
  fill: #ccc;
  margin-right: 0.25em;
`;

export default IssueIcon;
