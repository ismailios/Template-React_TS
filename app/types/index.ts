import { Reducer, Store } from 'redux';
import { RouterState } from 'connected-react-router';
import { Saga } from 'redux-saga';
import { SagaInjectionModes } from 'redux-injectors';

import { ContainerState as LanguageProviderState } from 'containers/LanguageProvider/types';
import { ContainerState as AppState } from 'containers/App/types';
import { ContainerState as HomeState } from 'containers/HomePage/types';
// [IMPORT NEW CONTAINERSTATE ABOVE] < Needed for generating containers seamlessly

export interface InjectedStore extends Store {
  injectedReducers: any;
  injectedSagas: any;
  runSaga(saga: Saga<any[]> | undefined, args: any | undefined): any;
}

export interface InjectReducerParams {
  key: keyof ApplicationRootState;
  reducer: Reducer<any, any>;
}

export interface InjectSagaParams {
  key: keyof ApplicationRootState;
  saga: Saga;
  mode?: SagaInjectionModes;
}

// Your root reducer type, which is your redux state types also
export interface ApplicationRootState {
  readonly router: RouterState;
  readonly global: AppState;
  readonly language: LanguageProviderState;
  readonly home: HomeState;
  // [INSERT NEW REDUCER KEY ABOVE] < Needed for generating containers seamlessly

  // for testing purposes
  readonly test: any;
}
