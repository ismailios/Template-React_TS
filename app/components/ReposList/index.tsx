import * as React from 'react';

import List from 'components/List';
import ListItem from 'components/ListItem';
import LoadingIndicator from 'components/LoadingIndicator';
import RepoListItem from 'containers/RepoListItem';
import { ContainerState, UserData } from '../../containers/App/types';

export type ReposListProps = Pick<ContainerState, 'loading' | 'error'> &
  Pick<UserData, 'repos'>;

function ReposList({ loading, error, repos }: ReposListProps) {
  if (loading) {
    return <List component={LoadingIndicator} />;
  }

  if (error !== false) {
    const ErrorComponent = () => (
      <ListItem item="Something went wrong, please try again!" />
    );
    return <List component={ErrorComponent} />;
  }
  if (repos !== undefined) {
    return <List items={repos} component={RepoListItem} />;
  }

  return null;
}

export default ReposList;
