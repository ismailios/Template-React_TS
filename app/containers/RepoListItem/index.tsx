/**
 * RepoListItem
 *
 * Lists the name and the issue count of a repository
 */

import React from 'react';
import { useSelector } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { FormattedNumber } from 'react-intl';

import { makeSelectCurrentUser } from 'containers/App/selectors';
import ListItem from 'components/ListItem';
import IssueIcon from './IssueIcon';
import IssueLink from './IssueLink';
import RepoLink from './RepoLink';
import Wrapper from './Wrapper';
import { Repo } from './types';

interface OwnProps {
  item: Repo;
}

interface DispatchProps {}

type Props = DispatchProps & OwnProps;
const stateSelector = createStructuredSelector({
  currentUser: makeSelectCurrentUser(),
});

export default function RepoListItem({ item }: Props) {
  const { currentUser } = useSelector(stateSelector);
  let nameprefix = '';

  // If the repository is owned by a different person than we got the data for
  // it's a fork and we should show the name of the owner
  if (item.owner.login !== currentUser) {
    nameprefix = `${item.owner.login}/`;
  }

  // Put together the content of the repository
  const content = (
    <Wrapper>
      <RepoLink href={item.html_url} target="_blank">
        {nameprefix + item.name}
      </RepoLink>
      <IssueLink href={`${item.html_url}/issues`} target="_blank">
        <IssueIcon />
        <FormattedNumber value={item.open_issues_count} />
      </IssueLink>
    </Wrapper>
  );

  // Render the content into a list item
  return <ListItem key={`repo-list-item-${item.full_name}`} item={content} />;
}
