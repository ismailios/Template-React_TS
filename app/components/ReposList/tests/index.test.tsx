import React from 'react';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

import ReposList from '../index';
import configureStore from '../../../configureStore';
import history from '../../../utils/history';
import { Repo } from '../../../containers/RepoListItem/types';

describe('<ReposList />', () => {
  it('should render the loading indicator when its loading', () => {
    const { container } = render(<ReposList loading />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render an error if loading failed', () => {
    const { queryByText } = render(
      <IntlProvider locale="en">
        <ReposList loading={false} error={{ message: 'Loading failed!' }} />
      </IntlProvider>,
    );
    expect(queryByText(/Something went wrong/)).toBeInTheDocument();
  });

  it('should render the repositories if loading was successful', () => {
    const initialState = {
      global: {
        currentUser: 'mxstbr',
        error: false,
        loading: false,
        userData: {
          repos: false,
        },
      },
    };
    const store = configureStore(initialState, history);
    const repos = [
      {
        owner: {
          login: 'mxstbr',
        },
        html_url: 'https://github.com/react-boilerplate/react-boilerplate',
        name: 'react-boilerplate',
        open_issues_count: 20,
        full_name: 'react-boilerplate/react-boilerplate',
      },
    ] as Repo[];
    const { container } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <ReposList repos={repos} error={false} loading={false} />
        </IntlProvider>
      </Provider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not render anything if nothing interesting is provided', () => {
    const { container } = render(
      <ReposList repos={undefined} error={false} loading={false} />,
    );

    expect(container).toBeEmpty();
  });
});
