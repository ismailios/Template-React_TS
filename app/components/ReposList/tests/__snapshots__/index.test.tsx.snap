// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReposList /> should render the loading indicator when its loading 1`] = `
<div
  class="sc-bwzfXH bycCir"
>
  <ul
    class="sc-bdVaJa kjaCOS"
  >
    <div
      class="sc-ifAKCX jcAqoe"
    >
      <div
        class="sc-iwsKbI dPWiyi"
      />
      <div
        class="sc-gZMcBi ebtRmY"
      />
      <div
        class="sc-gqjmRU cTErFR"
      />
      <div
        class="sc-VigVT dFmMlF"
      />
      <div
        class="sc-jTzLTM enuhco"
      />
      <div
        class="sc-fjdhpX jNqnwR"
      />
      <div
        class="sc-jzJRlG fEFjAE"
      />
      <div
        class="sc-cSHVUG Hyerz"
      />
      <div
        class="sc-kAzzGY hOJQGd"
      />
      <div
        class="sc-chPdSV kNRVZc"
      />
      <div
        class="sc-kgoBCf jBYbkL"
      />
      <div
        class="sc-kGXeez klHZt"
      />
    </div>
  </ul>
</div>
`;

exports[`<ReposList /> should render the repositories if loading was successful 1`] = `
<div
  class="sc-bwzfXH bycCir"
>
  <ul
    class="sc-bdVaJa kjaCOS"
  >
    <li
      class="sc-bxivhb cHNIkg"
    >
      <div
        class="sc-htpNat cdjzok"
      >
        <div
          class="sc-dnqmqq hIJpCG"
        >
          <a
            class="sc-bZQynM sc-htoDjs zeaOm"
            href="https://github.com/react-boilerplate/react-boilerplate"
            target="_blank"
          >
            react-boilerplate
          </a>
          <a
            class="sc-bZQynM sc-gzVnrw XFJuy"
            href="https://github.com/react-boilerplate/react-boilerplate/issues"
            target="_blank"
          >
            <svg
              class="sc-EHOje bDxpQi"
              height="1em"
              width="0.875em"
            >
              <path
                d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7S10.14 13.7 7 13.7 1.3 11.14 1.3 8s2.56-5.7 5.7-5.7m0-1.3C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7S10.86 1 7 1z m1 3H6v5h2V4z m0 6H6v2h2V10z"
              />
            </svg>
            <span>
              20
            </span>
          </a>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
